### PR TITLE
Interactive confusion matrix

### LINKED ISSUE #185: interactive plots
We've seen that most people use the same subset of plots; hence, adding more plots does not yield more usage. We must think about how to make the most popular plots more useful. I'll keep editing this issue as I do more research.

## [confusion matrix](https://sklearn-evaluation.ploomber.io/en/latest/api/plot.html#confusion-matrix)

- clicking on any quadrant in the confusion matrix could sample a few rows from such quadrant to the user can inspect them (looking at individual plots often provides good insights)
- alternatively, we might want to provide a "profile" of all the observations in such quadrant (e.g., mean/media/percentiles for each feature). This can help the user find patterns in the errors (e.g., the model is making false positives for houses with prices >1M)

## considerations

we can start building this step by step, and provide enhanced functionality to gauge interest. For example, we can offer a similar level of interactivity with an API like this:

```python
from sklearn_evaluation import plot

cm = plot.ConfusionMatrix.from_raw_data(y_true, y_pred)

# get indexes of false positives (same for false negatives, true positives, and true negatives)
idx_fp = cm.get_fp()

# get the observations
X_test_fp = X_test.iloc[idx_fp]

# I think pandas has  function like this
X_test_fp.summary()
````

If we se that our users like this `get_fp`, then we can work on providing the interactivity so they don't have to perform all this filtering and slicing manually.

## open questions

- can we use plotly or something similar? or do we need HTML/JS/CSS?

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes
Interactive confusion matrix plots.

## Issue ticket number and link
Closes #185 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [x] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://sklearn-evaluation--275.org.readthedocs.build/en/275/

### CODE REVIEW

**Comment by neelasha23** [comment]
Few clarifications before proceeding with the code:

 Currently I have added the code as a [user guide](https://sklearn-evaluation--275.org.readthedocs.build/en/275/classification/cm_interactive.html). (Will move this code to the `ConfusionMatrix` class once there's more clarity. Right now if we hover over the quadrants it displays a `Min` metric (dummy values).

1. `we could show some statistics (range, min, max, percentiles) for the data points in such quadrant` : How do we calculate these values? The plots are of `actual labels vs predicted labels`. Do we need to calculate statistics on the test features? 
2. Can we add a parameter `interactive=True/False` to the `from_raw_data` . This will call the interactive version of the plot.

@edublancas

**Comment by neelasha23** [comment]
Have added table next to the plot. The values change depend on which quadrant is clicked. It has some initial values which I'm looking at removing.
@edublancas

**Comment by edublancas** [comment]
great! I think let's implement this as `InteractiveConfusionMatrix` with the same structure as AbstractPlot (there will be some minor differences since we no longer have the `ax` and `figure` attributes, but most of the structure will be the same.

The signature can be something like:

```python
def from_raw_data(cls, y_true, y_pred, target_names=None, normalize=False, X_test=None):
    ....
```

The key difference here is `X_test`. If None, we just display the matrix, but if not None, we display two tables below the matrix:

1. randomly samples 3-5 observations that correspond to the clicked quadrant
2. computes some summary statistics from the columns

The current function we use for the other confusion matrix returns us the matrix directly, so we'll need to implement some logic to determine for each observation, which quadrant it belongs to.

We can assume that X_test will be a data frame, we can later decide what we do with numpy arrays (since they don't have column labels, it doesn't make much sense to display them in a table)

File: setup.py [hunk 1]
```diff
@@ -97,6 +97,7 @@
         # extracting injected parameters from notebooks
         "parso",
         'importlib-metadata;python_version<"3.8"',
+        "altair"
```
Comment (edublancas) [inline setup.py:100]: let's make this an optional dependency https://ploomber-core.readthedocs.io/en/latest/dependencies.html

**Comment by edublancas** [comment]
for the sample rows: can we display them as a pandas-likes table? the numbers with no column labels is a bit hard to read

> Still figuring out the histogram part and what data to exactly show there. 

I was thinking of simple summary statistics, like the ones [pandas.describe](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) shows

> In this case we can mandate the user to pass column names as well, and then internally convert numpy to dataframe.

Ok, so it can be:

```
from_raw_data(cls, y_true, y_pred, target_names=None, normalize=False, X_test=None, feature_names=None):
```

if X_test is a data frame, we grab them from there, if it's a numpy array, we require feature_names (or maybe if feature_names is None, we display feature 0, feature 1,...)

**Comment by neelasha23** [comment]
> for the sample rows: can we display them as a pandas-likes table? the numbers with no column labels is a bit hard to read

column labels are present (currently it shows Feature N_sampled, I'll change it to just the feature N). I have removed the spacing between columns. [This](https://altair-viz.github.io/gallery/scatter_linked_table.html) is the tutorial I followed. There is no support of tables in altair so this is a workaround . [Refer](https://github.com/altair-viz/altair/issues/1730)

> I was thinking of simple summary statistics, like the ones [pandas.describe](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html) shows

Added `mean`, `min`, `max` for now, will add more if this looks fine.

@edublancas

**Comment by neelasha23** [comment]
Have addressed the review suggestions. A few doubts:

1. Altair charts rendering is well integrated with jupyter notebook but not when we run a python script. Tried the methods mentioned [here](https://altair-viz.github.io/user_guide/display_frontends.html). So in the docstring of the `from_raw_method` the chart is being displayed using `.. raw:: html` directive. So everytime the code changes the new chart after running the example needs to be copied and saved in `src/sklearn-evaluation/assets/cm/` path as `metric_chart.html`. I'm not sure if this is the correct way as it's a manual process, but I'm still not sure how else to handle this.
2. If `X_test` contains too many columns the view would become very wide, in this case should users pass only the desired columns as a dataframe/ numpy array. Or should this be handled from our end?

@edublancas

**Comment by neelasha23** [comment]
Did you get a chance to review this?
@edublancas

**Comment by edublancas** [comment]
sorry for the delay

> Altair charts rendering is well integrated with jupyter notebook but not when we run a python script. Tried the altair_viewer method mentioned [here](https://altair-viz.github.io/user_guide/display_frontends.html), but it opens the chart in a new browser window during the docs build. So in the docstring of the from_raw_method the chart is being displayed using .. raw:: html directive. So everytime the code changes the new chart after running the example needs to be copied and saved in src/sklearn-evaluation/assets/cm/ path as metric_chart.html. I'm not sure if this is the correct way as it's a manual process, but I'm still not sure how else to handle this.

let's ignore this for now and just include a link from the docstring to a notebook tutorial so it correctly displays the interactive plot. we can create a new one under classification/interactive-plots

**Comment by edublancas** [comment]
> If X_test contains too many columns the view would become very wide, in this case should users pass only the desired columns as a dataframe/ numpy array. Or should this be handled from our end?

if there are more than X columns, I think let's just grab the top Y (not sure how many make sense, maybe 4-5?) and display a warning saying that only the top X columns are displayed and that users might subselect fewer columns if they want to control which ones to display

**Comment by neelasha23** [comment]
Fixed the above points
@edublancas

**Comment by tonykploomber** [comment]
In the preview doc
https://sklearn-evaluation--275.org.readthedocs.build/en/275/classification/cm_interactive.html

It seems the Observation and Statistics number are overlapping

Screenshot 2023-02-23 at 11 34 04 AM

**Comment by neelasha23** [comment]
> In the preview doc https://sklearn-evaluation--275.org.readthedocs.build/en/275/classification/cm_interactive.html
> 
> It seems the Observation and Statistics number are overlapping
> 
> Screenshot 2023-02-23 at 11 34 04 AM

Yes this is happening if you click on the table area. In this case all the quadrants are getting selected all at once.

**Comment by tonykploomber** [comment]
I also checked the altair doc
Seems there is no easy way to fix the issue when multiple selected mode.

We can check if there is better way to present table with multiple values

**Comment by edublancas** [comment]
is there a way to get back the original colors? and to highlight the selected one, maybe we can add a border or something else so it's clear which one is selected. I think with that we'd be ready to release!

looks like this:

image

although when double-clicking a quadrant, it gets the right colors (but triggers the overlap problem):

image



one thing that came to mind (let's see if it's not too much of a hassle): when testing the confusion matrix, I realized that it'd be useful to compare quadrants. for example, I might want to know how false positives for class 1 compare to false positives for class 2 by looking at the sample statistics. is there a way to allow a user to select two quadrants? for example, the first one with and a second one with a double-click or some other modifies like shift + click

then if two quadrants are selected, we could display the ratio of features: feature_1_from_quadrant_1 / feature_1_from_quadrant_2

**Comment by neelasha23** [comment]
> is there a way to get back the original colors? and to highlight the selected one, maybe we can add a border or something else so it's clear which one is selected. I think with that we'd be ready to release!

I don't see a `border` attribute in encodings. There is `opacity`, I tried it along with conditional colors but it's messing up the color.
The colors are currently selected based on a quadrant click like 
```
color = alt.condition(
            selection,
            "confusion_matrix:N",
            alt.value("lightgray"),
            scale=alt.Scale(scheme="oranges"),
        )
```

This condition has a format `(event_selection, color_if_true, color_if_false)

The only thing I can think of is changing the `lightgray` to some other color but that doesn't seem to solve the purpose.


>  when testing the confusion matrix, I realized that it'd be useful to compare quadrants.

There is a way to select multiple quadrants using `selection_multi` (Shift button) but it is displaying an overlapping chart (similar when we select all quadrants)
I think the we would somehow need to capture the event `single quadrant clicked` and `double quadrant clicked` in our script .
```
if single_quadrant_clicked:
  display tables
else if double_quadrant_clicked:
  display feature ratios
```

I'm not sure how to go about this. Would need some more time to experiment

File: setup.py [hunk 2]
```diff
@@ -27,6 +27,7 @@
     "ipywidgets",
     # notebook database example
     "jupysql",
+    "altair",
```
Comment (edublancas) [inline setup.py:30]: please add a comment explainign that altair is needed for interactive confusion matrix

File: setup.py [hunk 3]
```diff
@@ -46,14 +47,10 @@
     "pytest-cov",
     # TODO: update config so coveralls 3 works
     "coveralls<3",
+    "altair",
```
Comment (edublancas) [inline setup.py:50]: please add a comment explainign that altair is needed for interactive confusion matrix

**Review by edublancas — CHANGES_REQUESTED** [review]
looks good. only minor changes requested

I opened https://github.com/ploomber/sklearn-evaluation/issues/302 and https://github.com/ploomber/sklearn-evaluation/issues/301 to tackle the issues later, this version is good enough to make a release

**Comment by idomic** [comment]
@neelasha23 anything else is pending here?

**Comment by neelasha23** [comment]
This is done, I have resolved all the comments. 
@idomic

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
Woohoo! 🎉

**Comment by idomic** [comment]
Nice job @neelasha23 !

**Pull request merged**